src/main/java/io/netty/handler/codec/compression/BrotliDecoder.java

```diff
@@ -0,0 +1,111 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.codec.compression;
+
+import com.nixxcode.jvmbrotli.common.BrotliLoader;
+import com.nixxcode.jvmbrotli.dec.BrotliInputStream;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.Arrays;
+import java.util.List;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufInputStream;
+import io.netty.buffer.ByteBufOutputStream;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.handler.codec.ByteToMessageDecoder;
+import io.netty.util.internal.logging.InternalLogger;
+import io.netty.util.internal.logging.InternalLoggerFactory;
+
+/**
+ * Decompress a {@link ByteBuf} using the inflate algorithm.
+ */
+public class BrotliDecoder extends ByteToMessageDecoder {
+
+  private static final InternalLogger log =
+      InternalLoggerFactory.getInstance(BrotliDecoder.class);
+
+  static {
+    BrotliLoader.isBrotliAvailable();
+  }
+
+  /*
+  For how this value is derived, please see: `BROTLI_MAX_NUMBER_OF_BLOCK_TYPES` in these docs:
+     - https://github.com/google/brotli/blob/master/c/common/constants.h
+     - https://tools.ietf.org/html/draft-vandevenne-shared-brotli-format-01
+   */
+  private static int BROTLI_MAX_NUMBER_OF_BLOCK_TYPES = 256;
+
+  public BrotliDecoder() {
+  }
+
+  public static byte[] decompress(byte[] compressedArray) throws IOException {
+    if(compressedArray == null) {
+      return null;
+    }
+
+    ByteArrayOutputStream out;
+    try (BrotliInputStream is = new BrotliInputStream(new ByteArrayInputStream(compressedArray))) {
+      out = new ByteArrayOutputStream();
+      if (!decompress(out, is)) {
+        return null;
+      }
+    } catch (IOException e) {
+      log.error("Unhandled exception when decompressing brotli", e);
+      throw e;
+    }
+    return out.toByteArray();
+  }
+
+  @Override
+  protected void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) throws Exception {
+    /*
+       use in.alloc().buffer() instead of Unpooled.buffer() as best practice.
+       See: https://github.com/netty/netty/wiki/New-and-noteworthy-in-4.0#pooled-buffers
+    */
+    try (ByteBufOutputStream output = new ByteBufOutputStream(in.alloc().buffer())) {
+      try (BrotliInputStream brotliInputStream = new BrotliInputStream(new ByteBufInputStream(in))) {
+        if (!decompress(output, brotliInputStream)) return;
+      } catch (IOException e) {
+        log.error("Unhandled exception when decompressing brotli", e);
+        throw e;
+      }
+      out.add(output.buffer());
+    }
+  }
```

Comment:
```suggestion
  @Override
  protected void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) throws Exception {
    /*
       use in.alloc().buffer() instead of Unpooled.buffer() as best practice.
       See: https://github.com/netty/netty/wiki/New-and-noteworthy-in-4.0#pooled-buffers
    */
    boolean success = true;
    ByteBuf outBuffer = null;
    try (ByteBufOutputStream output = new ByteBufOutputStream(in.alloc().buffer())) {
      try (ByteBufInputStream bbin = new ByteBufInputStream(in)) {
        try (BrotliInputStream brotliInputStream = new BrotliInputStream(bbin)) {
          success = decompress(output, brotliInputStream);
          if (!success) {
            return;
          }
        } catch (IOException e) {
          log.error("Unhandled exception when decompressing brotli", e);
          throw e;
        }
      }

      outBuffer = output.buffer();
      if (outBuffer.isReadable()) {
        out.add(outBuffer);
      } else {
        success = false;
      }
    } finally {
      if (!success) {
        if (outBuffer != null) {
          outBuffer.release();
        }
      }
    }
  }
```

Reply:
this implementation follows practices established by other implementations of the decoder i found in the netty source files

Reply:
a buffer should be released if the decompression was unsuccessful. otherwise we have a memory leak

Reply:
reading properties of an output buffer should be checked. if they are unfit, the buffer is released.